openwpm_utils/s3.py

```diff
@@ -46,7 +57,22 @@ def read_table(self, table_name, columns=None):
         """
         table = self._sql_context.read.parquet(self._s3_table_loc % table_name)
         if columns is not None:
-            return table.select(columns)
+            table = table.select(columns)
+        if mode == "all":
+            return table
+        if mode == "failed":
```

Comment:
since joins can be quite expensive in spark, I'm wondering whether it will be more efficient to run a short operation on the crawl_history table that collects out a set of failed visit ids to the driver process, and then simply filter the future tables by that set of ids. I suspect the set of failed visits will be relatively small for each crawl? For example:

```python
table.where(~F.col("visit_id").isin(self._failed_visit_ids.union(self._incomplete_visit_ids)))
```

where `self._failed_visit_ids` and `self._incomplete_visit_ids` are sets stored in memory in the driver process.

Would you mind to do a comparison between the two approaches? I'll admit I'm not well versed in _how_ the cluster handles joins vs a simple filter like this, but I think it's worth testing because it will impact all future analyses run on the data loaded through this function. It might be worth paying the one-time cost to collect the bad visits from the crawl_history table at the start if filtering ends up being much faster than the joining.

Reply:
I tried this approach for the JS table and the results seem to indicate that a join is still more efficient than reducing this to a list.  
Creating the dataset object and loading the JS table: 60s
Creating the set of all bad visit ids with the dataset object already existing: 23s
Loading and filtering the JS table with the set: 120s
So I think we should stick with the current implementation

Reply:
I should have done collect or some other operations

Reply:
Our current implementation is set based and I don't want to change it for now, since we are hoping to soon do the filtering in a preprocessing step.

Reply:
From what I remember both approaches were similar in speed and I don't have a enough context to just rewrite this.
I propose just leaving it like it is rn and focusing on preprocessing the data, so that these selections can be made by specifying a different path.